File: Python3/328. Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        odds = head
        first_odd = head
        evens = head.next
        first_even = head.next
        head = evens.next
        count = 1
        while head:
            if count == 1:
                odds.next = head
                odds = odds.next
                count = 0
            else:
                evens.next = head
                evens = evens.next
                count = 1
            head = head.next
        evens.next = None # 防止出现环
        odds.next = first_even
        return first_odd

File: Golang/328. Odd Even Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func oddEvenList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    
    odd := head
    first_odd := head
    even := head.Next
    first_even := head.Next
    head = first_even.Next
    count := 1
    for head != nil {
        if count == 1 {
            odd.Next = head
            odd = odd.Next
            count = 0
        } else {
            even.Next = head
            even = even.Next
            count = 1
        }
        head = head.Next
    }
    even.Next = nil
    odd.Next = first_even
    return first_odd
}
